libraries/pybamm-developer/01_ode.md

```diff
@@ -0,0 +1,341 @@
+3. `initial_conditions` - a python dictionary of the initial conditions
+4. `variables` - a python dictionary of the output variables
+
+As an example, lets define a simple model for exponential decay with a single state variable $x$ and a single parameter $a$:
+
+```python
+x = pybamm.Variable("x")
+a = pybamm.Parameter("a")
+
+model = pybamm.BaseModel("exponential decay")
+model.rhs = {x: -a * x}
+model.initial_conditions = {x: 1}
+model.variables = {"x": x}
+```
+
+::::challenge{id="ode-model" title="Define the reservoir model"}
+
+Now we have all the pieces we need to define the reservoir model. Define the
+model using the parameters and variables you defined earlier.
+
+:::solution
+```python
+model = pybamm.BaseModel("reservoir model")
+model.rhs[x_n] = -i / Q_n
+model.initial_conditions[x_n] = x_n_0
+model.rhs[x_p] = i / Q_p
+model.initial_conditions[x_p] = x_p_0
+
+model.variables["Voltage [V]"] = U_p - U_n -  i * R
+model.variables["Negative electrode stochiometry"] = x_n
+model.variables["Positive electrode stochiometry"] = x_p
+```
+
+Note that we have defined the dictionaries in a slightly different way to the
+example above, this is just to show that there are multiple ways to define these
+dictionaries using Python.
+:::
+::::
+
+
+## PyBaMM expressions
+
+It is worth pausing here and discussing the concept of an "expression" in
+PyBaMM. Notice that in the model definition we have used expressions like `-i /
+Q_n` and `U_p - U_n -  i * R`. These expressions do not evaluate to a single
+value like similar expressions involving Python variables would. Instead, they
+are symbolic expressions that represent the mathematical equation.
+
+The fundamental building blocks of a PyBaMM model are the expressions, either
+those for the RHS rate equations, the algebraic equations of the expression for
+the output variables such as $V(t)$. PyBaMM encodes each of these expressions
+using an "expression tree", which is a tree of operations (e.g. addition,
+multiplication, etc.), variables (e.g. $x_n$, $x_p$, $I(t)$, etc.), and
+functions (e.g. $\exp$, $\sin$, etc.), which can be evaluated at any point in
+time.
+
+As an example, lets consider the expression for $\frac{dx_n}{dt}$, given by `-i
+/ Q_n`. We can visualise the expression tree for this expression using the
+`visualise` method:
+
+```python
+model.rhs[x_n].visualise("x_n_rhs.png")
+```
+
+This will create a file called `x_n_rhs.png` in the current directory, which you
+can open to see the expression tree, which will look like this:
+
+![x_n_rhs](images/x_n_rhs.png)
+
+You can also print the expression as a string using the `print` method:
+
+```python
+print(model.rhs[x_n])
+```
+
+```
+-Current function [A] / Negative electrode capacity [A.h]
+```
+
+So you can see that the expression tree is a symbolic representation of the
+mathematical equation, which can then be later on used by the PyBaMM solvers to
+solve the model equations over time.
```

Comment:
Here can we show how to use `children` to access subsections of the expression tree, which can be useful for debugging?